source/language/pulses.rst

```diff
@@ -92,3 +83,63 @@ circuit. Most of the time symbols in the ``defcal`` table will also have
 corresponding ``gate`` definitions. However, if a user provides a ``defcal`` for a symbol
 without a corresponding ``gate``, then we treat such operations like the ``opaque`` gates
 of prior versions of OpenQASM.
+
+Restrictions on defcal bodies
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The contents of ``defcal`` bodies are subject to some restrictions. This is
+similar to how boxed expressions work.
+
+- They must have a definite length, regardless of the parameters passed in or
+  the state of the system when called. This allows the compiler to properly
+  resolve ``lengthof(...)`` calls and allowing for optimizations.
+- No control flow is allowed within a ``defcal`` block (with one exception,
+  see below). This follows from the definite length requirement.
```

Comment:
We need a better way to express this constraint that doesn't feel like such a kludge. I don't have a good suggestion for improvement right now, though.